### ISSUE TITLE
Gracefully handle failure case of JSON.parse

### ISSUE THREAD

**Issue by AJCStriker** [dodekeract/spirit-body#6]
It seems that providing invalid JSON to the parser would trigger an error but the error would not be reported correctly to the parent handler.

Maybe I missed something but it might be worth adding a try / catch around the JSON parser and rejecting the promise in the event of a failure. Ideally to keep in the spirit of spirit the error should just be returned so that a parent error handler could cleanly report the error.

Is this issue present and if so would you accept a pull request to add the handling and the relevant test case?

**Comment by dodekeract** [comment]
Yes, I think this is an issue. I'd gladly accept a PR for it or do it myself when I find some time.

Any idea how this should be properly handled? `try {} catch () {}` would work to get the error, but how should it be passed to the next middleware? Maybe just not set body or set a `bodyError` property?

**Comment by AJCStriker** [comment]
Well if we reject the promise with an error then any middleware higher up the chain would be able to catch it with: 
```
try { await handler(r) } catch (e) {}
```

We could perhaps add a flag to the options that defaults to true so the library returns an error response to the client by default but can throw the error up the chain if the user is a bit more advanced?

**Comment by dodekeract** [comment]
> Library returns an error response

Can we actually do this from the scope of this middleware? Would be nice to have for sure (as long as it's optional).

**Comment by AJCStriker** [comment]
As far as I have understood Spirit's method of operation I think we could.

My reasoning is that spirit middleware as I understand it is basically functional composition of the middleware around the user provided function.

By resolving the promise with a Response type directly and not calling the subsequent handler we would effectively curtail execution to within this middleware and begin returning control back up the stack.

TL;DR: I think that by calling the following we could return a nicely formatted error in the event of an incompatible string being provided.
```
resolve({ 
   status: 400, 
   body: JSON.stringify({
      message: "An error occurred."
   })}
```

**Comment by dodekeract** [comment]
@AJCStriker Just wanted to let you know, I'll start working on this. I'm assuming you haven't implemented anything yet, right?

Otherwise we may work both on the same stuff which is a stupid thing to do. :smile:

**Comment by AJCStriker** [comment]
@dodekeract I had done some exploratory stuff around throwing the errors but hadn't forked anything yet. Feel free to implement if you want!

**Comment by dodekeract** [comment]
@AJCStriker This should be working now. Can you take a look at it and let me know if you'd like to change something? The project is still pretty new and if we want to make breaking changes we should do it as soon as we can. :smile:
